### PR TITLE
change the behavior of `ui.cursor.insert` and `ui.cursor.select` when `ui.cursor.primary` is present

### LINKED ISSUE #1337: `ui.cursor.primary` always overrides `ui.cursor.insert`
If `ui.cursor.primary` is set in a theme, it always overrides e.g. `ui.cursor.insert`.

E.g. currently in gruvbox theme one can't tell where is the cursor in a selection because it has the same color as the selection:
image

If I change the `ui.cursor.primary` value of the theme file, the issue is solved:
image

However, I also would like to give `ui.cursor.insert` a different color, so I know if I am in insert mode, but currently it is not possible

### PR DESCRIPTION
**Pull request by oati**
Fixes #1337

Previously, `ui.cursor.insert` and `ui.cursor.select` only applied to non-primary cursors when `ui.cursor.primary` was present.
This PR changes this behavior to make `ui.cursor.insert` and `ui.cursor.select` only affect the primary cursor when `ui.cursor.primary` is set.

### CODE REVIEW

**Comment by oati** [comment]
The motivation is that `ui.cursor.{insert,select}` are currently useless when `ui.cursor.primary` is set. The effect of `ui.cursor.{insert,select}` is only visible in non-primary cursors.

This change would make `ui.cursor.{insert,select}` "play better" with `ui.cursor.primary`, making the behavior closer to how cursor shapes work.

Another option to consider is adding the options `ui.cursor.primary.{insert,select}` or `ui.cursor.{insert,select}.primary` to allow more complex theming.

File: helix-term/src/ui/editor.rs [hunk 1]
```diff
@@ -288,16 +288,25 @@ impl EditorView {
...
+        } else {
+            mode_cursor_scope.unwrap_or(base_cursor_scope)
+        };
+
+        let primary_cursor_scope =
+            if let Some(base_primary_cursor_scope) = theme.find_scope_index("ui.cursor.primary") {
+                mode_cursor_scope.unwrap_or(base_primary_cursor_scope)
+            } else {
+                cursor_scope
+            };
```
Comment (archseer) [inline helix-term/src/ui/editor.rs:308]: I'm a bit confused by the conditionals here.

**Review by archseer — COMMENTED** [review]
> Another option to consider is adding the options ui.cursor.{insert,select}.primary to allow more complex theming.

This seems more reasonable to me but it's tricky. What would the fallback there be? `ui.cursor.primary`? Or `ui.cursor.insert`

File: helix-term/src/ui/editor.rs (hunk 1 above)
Comment (oati) [reply helix-term/src/ui/editor.rs:308]: this chooses between `mode_cursor_scope` and `base_cursor_scope`/`base_primary_cursor_scope` depending on whether `ui.cursor.primary` is present.

I'm considering making a new PR that cleans up the code to make all of the defaults and fallbacks clearer, and adding code for `ui.cursor.{insert,select}.primary` over it.

**Comment by oati** [comment]
> This seems more reasonable to me but it's tricky. What would the fallback there be? `ui.cursor.primary`? Or `ui.cursor.insert`

The current behavior chooses `ui.cursor.primary`. Ideally, there shouldn't be any situation where `ui.cursor.primary` and `ui.cursor.insert` are both present without `ui.cursor.insert.primary`.

One argument for making `ui.cursor.primary` first is that people *might* want a consistent primary cursor color with varying cursor shape, while having colors for non-primary cursors.

Falling back to `ui.cursor.insert` would make you explicitly specify the primary cursor for each mode, which might be better for forcing themes to make sense (i.e. avoid situations where only the non-primary cursors change colors between modes).

**Comment by archseer** [comment]
Replaced by #5130

**Pull request closed without merging**